src/driver/alloy.rs

```diff
@@ -0,0 +1,223 @@
+    fn from(inner: EVMError<Db::Error>) -> Self {
+        Self::EVMError { inner }
+    }
+}
+
+impl<Db: revm::Database> std::fmt::Debug for BundleError<Db> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::TimestampOutOfRange => write!(f, "TimestampOutOfRange"),
+            Self::BlockNumberMismatch => write!(f, "BlockNumberMismatch"),
+            Self::BundleReverted => write!(f, "BundleReverted"),
+            Self::TransactionDecodingError(e) => write!(f, "TransactionDecodingError({:?})", e),
+            Self::EVMError { .. } => write!(f, "EVMError"),
+        }
+    }
+}
+
+/// A block filler for the bundle, used to fill in the block data specified for the bundle.
+#[derive(Clone, Debug)]
+struct BundleBlockFiller {
+    pub block_number: BlockNumberOrTag,
+    pub timestamp: Option<u64>,
+    pub gas_limit: Option<u64>,
+    pub difficulty: Option<U256>,
+    pub base_fee: Option<u128>,
+}
+
+impl Block for BundleBlockFiller {
+    fn fill_block_env(&self, block_env: &mut revm::primitives::BlockEnv) {
+        if let Some(timestamp) = self.timestamp {
+            block_env.timestamp = U256::from(timestamp);
+        }
+        if let Some(gas_limit) = self.gas_limit {
+            block_env.gas_limit = U256::from(gas_limit);
+        }
+        if let Some(difficulty) = self.difficulty {
+            block_env.difficulty = difficulty;
+        }
+        if let Some(base_fee) = self.base_fee {
+            block_env.basefee = U256::from(base_fee);
+        }
+        if let Some(block_number) = self.block_number.as_number() {
+            block_env.number = U256::from(block_number);
+        }
+    }
+}
+
+impl From<EthCallBundle> for BundleBlockFiller {
+    fn from(bundle: EthCallBundle) -> Self {
+        Self {
+            block_number: bundle.state_block_number,
+            timestamp: bundle.timestamp,
+            gas_limit: bundle.gas_limit,
+            difficulty: bundle.difficulty,
+            base_fee: bundle.base_fee,
+        }
+    }
+}
+
+impl<Ext> BundleDriver<Ext> for EthCallBundle {
+    type Error<Db: revm::Database> = BundleError<Db>;
+
+    fn run_bundle<'a, Db: revm::Database + revm::DatabaseCommit>(
+        &mut self,
+        trevm: crate::EvmNeedsTx<'a, Ext, Db>,
+    ) -> DriveBundleResult<'a, Ext, Db, Self> {
+        // 1. Check if the block we're in is valid for this bundle. Both must match
+        if trevm.inner().block().number.to::<u64>() != self.block_number {
+            return Err(trevm.errored(BundleError::BlockNumberMismatch));
+        }
+
+        let bundle_filler = BundleBlockFiller::from(self.clone());
+
+        let run_result = trevm.try_with_block(&bundle_filler, |trevm| {
+            let mut trevm = trevm;
+
+            for raw_tx in self.txs.clone().into_iter() {
+                let tx = TxEnvelope::decode_2718(&mut raw_tx.to_vec().as_slice());
+
+                let tx = match tx {
```

Comment:
opt question should tx decoding be done greedily? I think so. i.e. we deser all txns before we run any through the evm